README.md
# t3m Git Branching System #

[version_shield]: https://img.shields.io/badge/version-1.0.0--rc01-blue.svg
[release_page]: https://github.com/mfederczuk/t3m/releases/tag/v1.0.0-rc01 "Release v1.0.0-rc01"
[![version: 1.0.0-rc01][version_shield]][release_page]
[![Changelog](https://img.shields.io/badge/-Changelog-blue.svg)](CHANGELOG.md "Changelog")

## About ##

**t3m** is a **Git** branching system that was based off of and inspired by other popular branching models and systems
like [Vincent Driessen's "A successful Git branching model"] (**git-flow**) and **[GitHub Flow]**.

This repository holds the specification document for **t3m** - read it [here](docs/index.md).  
Without the **CC BY-SA 4.0** section, the entire document takes about 20 minutes to read.

[Vincent Driessen's "A successful Git branching model"]: https://nvie.com/posts/a-successful-git-branching-model "A successful Git branching model &raquo; nvie.com"
[GitHub Flow]: https://guides.github.com/introduction/flow/ "Understanding the GitHub flow &middot; GitHub Guides"

## Contributing ##

Read through the [t3m Contribution Guidelines](CONTRIBUTING.md) if you want to contribute to this project.

## License ##

The **t3m** specification document is licensed under [**CC BY-SA 4.0**](licenses/CC-BY-SA-4.0.txt).  
For more information about copying and licensing, see the [COPYING.txt](COPYING.txt) file.


CHANGELOG.md
<!-- markdownlint-disable no-duplicate-heading -->

# Changelog #

All notable changes to this project will be documented in this file.
The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

# [Unreleased] #

## Added ##

* [RFC 2119] notice and the keywords are in all uppercase
* Possible `dev` alias for the development branch

[Unreleased]: https://github.com/mfederczuk/t3m/compare/v1.0.0-rc01..v1
[RFC 2119]: https://datatracker.ietf.org/doc/html/rfc2119

# [v1.0.0-rc01] - 2021-05-17 #

Initial Release

[v1.0.0-rc01]: https://github.com/mfederczuk/t3m/releases/tag/v1.0.0-rc01


check_rules.ts
/*
 * Script to search for writing & formatting violations and other mistakes.
 * Copyright (C) 2021  Michael Federczuk
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import escapeStringRegexp from "escape-string-regexp";

function wordPatterns(firstWordsString: (string | RegExp), ...restWordsStrings: (string | RegExp)[]): RegExp[] {
	return [
		firstWordsString,
		...restWordsStrings
	].map((wordsString) => {
		let words: string[];
		let flags: string;

		if(wordsString instanceof RegExp) {
			words = wordsString.source.split(/\s+/);
			flags = wordsString.flags;
		} else {
			words = wordsString.split(/\s+/).map(escapeStringRegexp);
			flags = "im";
		}

		return new RegExp(`\\b${words.join("\\s+")}\\b`, flags);
	});
}

/**
 * These are the writing and formatting rules/guidelines for the t3m specifications.
 */
const rules: {
	readonly [description: string]: (RegExp | RegExp[]);
} = {
	// == writing guidelines ======================================================================================== //

	"do not refer to the writer": wordPatterns(/i(?!\.e\.:)/im, "my", "mine", "i'm", "myself", "me"),
	"do not refer to the reader": wordPatterns("you", "your", "yours", "you're", "yourself"),

	"do not use \"for example\" (except at the beginning of sentences)": wordPatterns(/(?<!^)for example/im),
	"do not use \"that means\" (except at the beginning of sentences)":  wordPatterns(/(?<!^)that means/im),

	"do not use \"forbidden\"":                      wordPatterns("forbidden"),
	"do not use \"shall\" or \"shall not\"":         wordPatterns("shall", "shall not"),
	"do not use \"can\", \"cannot\" or \"can not\"": wordPatterns("can", "cannot", "can not", "can't"),

	"write ellipsis with exactly three consecutive periods": /\.{4,}/m,
	"write ellipsis without whitespace between the periods": /\.(\s+\.)+/m,

	"do not write whitespace before punctuation": /\s+[,;]/m,
	"write whitespace after punctuation":         /[,;]\S/m,

	"do not use \"etc.\" in parenthesis":         /etc\s*\.(?=\s*\))/im,
	"do not use ellipsis outside of parenthesis": /\.{3}(?!\s*\))/m,

	// === formatting guidelines ==================================================================================== //

	"do not use a backslash to terminate lines": /\\$/m,

	"no trailing whitespace after two spaces":     /(?<=  )\s+$/m,
	"no singular whitespace character at the end": /(?<=\S+) $/m,
	// TODO no trailing whitespace that isn't space
	"no lines just containing whitespace":         /^\s+$/m,

	"no letters or digits after column 120": /^(?!\[).{120}((\s+\w+)+|\w+(\s+\w+)*)/,
	"no lines longer than 200":              /^.{200}.+$/m,

	"do not use equals characters for headings":             /^=+$/m,
	"headings are formatted with octothorpes on both sides": /^#[^#]*?(?!#)$/m, // this isn't really optimal, but
	                                                                            // markdownlint should catch those
	                                                                            // errors

	"no line with just one or two dashes":                  /^--?$/m,
	"horizontal separators must have exactly three dashes": /^-{4,}$/m,

	"strong emphasis is formatted with two asterisks":                             /_\*|\*_|__/,
	"explicitly called out names and inline quotes are formatted with underlines": /\*"[^"]*"\*/,

	"files must only contain printable characters and spaces": /[\0-\x09\x0B-\x1F\x7F]/m,

	"files must always end with exactly one newline character": /[^\n]$/,

	// === other ==================================================================================================== //

	"spellcheck": wordPatterns("branche") // common mistake when trying to turn plural "branches" to singular "branch"
};

export default rules;


check-src/main.ts
import { stderr as chalk } from "chalk";
import { Parser } from "commonmark";
import { existsSync, readFileSync } from "fs";
import { LintError as MarkdownLintError, sync as markdownlintSync } from "markdownlint";
import { inspect } from "util";
import { Problem } from "./problem";
import { reportT3mProblems } from "./reportT3mProblems";
import { readdirRecursiveSync } from "./utils";

const inputDirectoryPath = process.argv[2];

if(!existsSync(inputDirectoryPath)) {
	console.error(`${inspect(inputDirectoryPath)} does not exist. Aborting.`);
	process.exit();
}

const filepaths = readdirRecursiveSync(inputDirectoryPath)
	.filter((filepath) => filepath.endsWith(".md"))
	.filter((filepath) => filepath.endsWith("test.ignore.md"))
	.filter((filepath) => !filepath.endsWith("cc-by-sa-4.0.md"));

const markdownlintResults = markdownlintSync({
	files: filepaths,
	config: { "line-length": false }
});

const parser = new Parser();

function mapMarkdownlintErrorToProblem(
	markdownlintError: MarkdownLintError,
	filePath: string,
	fileLines: readonly string[]
): Problem {

	let description = `${markdownlintError.ruleNames.join("/")} ${markdownlintError.ruleDescription}`;
	if(markdownlintError.errorDetail) {
		description += ` [${markdownlintError.errorDetail}]`;
	}

	const lineContents = fileLines[markdownlintError.lineNumber - 1];

	let begin = 1, length = lineContents.length;
	if(markdownlintError.errorRange !== null) {
		begin = markdownlintError.errorRange[0];
		length = markdownlintError.errorRange[1];
	}

	return {
		origin: "markdownlint",
		filepath: filePath,
		message: description,
		lines: [{
			nr: markdownlintError.lineNumber,
			contents: lineContents,
			rangeBegin: begin,
			rangeLength: length
		}]
	};
}

const problems = filepaths
	.flatMap((filepath) => {
		const fileContents = readFileSync(filepath).toString();
		const fileLines: readonly string[] = fileContents.split("\n");

		const t3mProblems = reportT3mProblems(filepath, fileLines, parser.parse(fileContents));

		const markdownlintProblems = markdownlintResults[filepath]
			.map((markdownlintError) => mapMarkdownlintErrorToProblem(markdownlintError, filepath, fileLines));

		return [...t3mProblems/* , ...markdownlintProblems */];
	});

function printProblem(problem: Problem) {
	const fileInfoStyle = chalk.bold;
	const errorNonWsStyle = chalk.red.bold;
	const errorWsStyle = chalk.red.bgRed;

	const firstLine = problem.lines[0];

	let msg = "";

	msg += fileInfoStyle(`${problem.filepath}:${firstLine.nr}:${firstLine.rangeBegin}:`);
	msg += ` (${problem.origin}) ${problem.message}`;

	problem.lines
		.forEach((line) => {
			const rangeEnd = (line.rangeBegin - 1) + line.rangeLength;

			const contentsBeforeRange = line.contents.substring(0, line.rangeBegin - 1);
			const contentsInRange = line.contents.substring(line.rangeBegin - 1, rangeEnd);
			const contentsAfterRange = line.contents.substring(rangeEnd);

			msg += `\n${line.nr.toString().padStart(5, " ")} | `;

			const pattern = (() => {
				if(contentsBeforeRange.length === 0 && contentsAfterRange.length === 0) {
					return /^(?<leadingWs>\s*)(?<inside>.*?)(?<trailingWs>\s*)$/;
				} else if(contentsBeforeRange.length === 0) {
					return /^(?<leadingWs>\s*)(?<inside>.*)$/;
				} else if(contentsAfterRange.length === 0) {
					return /^(?<inside>.*?)(?<trailingWs>\s*)$/;
				} else {
					return /(?<inside>.*)/;
				}
			})();

			msg += contentsBeforeRange;

			const match = contentsInRange.match(pattern)!;
			msg += errorWsStyle(match.groups!["leadingWs"] ?? "");
			msg += errorNonWsStyle(match.groups!["inside"] ?? "");
			msg += errorWsStyle(match.groups!["trailingWs"] ?? "");

			msg += contentsAfterRange;

			msg += "\n" + " ".repeat(8 + contentsBeforeRange.length);
			if(contentsInRange.length === 0) {
				msg += errorNonWsStyle("^");
			} else {
				msg += errorNonWsStyle("~".repeat(contentsInRange.length));
			}
		});

	console.error(msg);
}

if(problems.length > 0) {
	process.exitCode = 32;

	printProblem(problems[0]);

	problems
		.slice(1)
		.forEach((problem) => {
			console.error();
			printProblem(problem);
		});
}


check-src/reportT3mProblems.ts
import assert from "assert";
import { Node as CommonMarkNode, NodeType, NodeWalkingStep } from "commonmark";
import { inspect } from "util";
import { Problem } from "./problem";
import { T3mProblemReporter } from "./t3mProblemReporter";

// TODO

function wordPatterns(...words: [string, ...string[]]) {

}

function walkFiltered(
	documentNode: CommonMarkNode,
	nodeTypeFilter: NodeType,
	stepHandler: (stepNode: CommonMarkNode) => void
) {

	const walker = documentNode.walker();
	let step: (NodeWalkingStep | null);

	while((step = walker.next()) !== null) {
		if(step.node.type !== nodeTypeFilter) {
			continue;
		}

		stepHandler(step.node);
	}
}

export function reportT3mProblems(
	filePath: string,
	fileLines: readonly string[],
	documentNode: CommonMarkNode
): Problem[] {

	const problemReporter = new T3mProblemReporter();

	walkFiltered(documentNode, "text", (node) => {
		console.log(inspect(node.sourcepos, false, null, true));
		// console.log(node.prev);
		console.log();
	});

	walkFiltered(documentNode, "thematic_break", (node) => {
		const [[startLine, startColumn], [endLine, endColumn]] = node.sourcepos;

		assert(startLine === endLine, "Start line and end line of thematic break are not the same");

		if(startColumn > 1) {
			problemReporter
				.newProblem()
				.withMessage("Horizontal separators must begin at column nr 1")
				.atLine(startLine).inRange({ begin: 1, length: startColumn - 1 });
		}

		const thematicBreak = fileLines[startLine - 1].substring(startColumn - 1, endColumn);

		if(thematicBreak !== "---") {
			problemReporter
				.newProblem()
				.withMessage("Horizontal separators must consist of exactly three dashes")
				.atLine(startLine).inRange({ begin: startColumn, length: endColumn - startColumn + 1 });
		}
	});

	return problemReporter.retrieveProblems(filePath, fileLines);
}


check-src/t3mProblemReporter.ts
import { Problem, ProblemLine } from "./problem";
import { NonEmptyArray } from "./utils";

export interface T3mProblemReportCreatorLineRange {
	inRange(range: { begin: number, length: number; }): T3mProblemReportCreatorLine;
}

export interface T3mProblemReportCreatorLine {
	atLine(lineNr: number): T3mProblemReportCreatorLineRange;
}

export interface T3mProblemReportCreator {
	withMessage(message: string): T3mProblemReportCreatorLine;
}

interface TemporaryProblemLine {
	readonly nr: number;
	range?: {
		readonly begin: number;
		readonly length: number;
	};
}

interface TemporaryProblem {
	message?: string;
	lines: TemporaryProblemLine[];
}

export class T3mProblemReporter {
	#problems: TemporaryProblem[] = [];

	newProblem(): T3mProblemReportCreator {
		const temporaryProblem: TemporaryProblem = { lines: [] };

		this.#problems.push(temporaryProblem);

		return {
			withMessage(message: string): T3mProblemReportCreatorLine {
				temporaryProblem.message = message;

				const lineCreator: T3mProblemReportCreatorLine = {
					atLine(lineNr: number): T3mProblemReportCreatorLineRange {
						const temporaryLine: TemporaryProblemLine = { nr: lineNr };

						temporaryProblem.lines.push(temporaryLine);

						return {
							inRange(range: { begin: number, length: number; }): T3mProblemReportCreatorLine {
								temporaryLine.range = range;
								return lineCreator;
							}
						};
					}
				};

				return lineCreator;
			}
		};
	}

	retrieveProblems(filePath: string, fileLines: readonly string[]): Problem[] {
		return this.#problems
			.map((temporaryProblem): Problem => {
				if(temporaryProblem.message === undefined) {
					throw new Error("Problem without message");
				}

				if(temporaryProblem.lines.length === 0) {
					throw new Error("Problem without lines");
				}

				const lines: readonly ProblemLine[] = temporaryProblem.lines
					.map((temporaryProblemLine) => {
						if(temporaryProblemLine.range === undefined) {
							throw new Error("Problem line without range");
						}

						return {
							nr: temporaryProblemLine.nr,
							contents: fileLines[temporaryProblemLine.nr - 1],
							rangeBegin: temporaryProblemLine.range.begin,
							rangeLength: temporaryProblemLine.range.length
						};
					});

				return {
					origin: "t3m",
					filepath: filePath,
					message: temporaryProblem.message,
					lines: NonEmptyArray(lines)
				};
			});
	}
}


check.ts
/*
 * Script to search for writing & formatting violations and other mistakes.
 * Copyright (C) 2021  Michael Federczuk
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <https://www.gnu.org/licenses/>.
 */

import chalk from "chalk";
import { PathLike, readdirSync, readFileSync, statSync } from "fs";
import markdownlint from "markdownlint";
import { normalize as normalizePath } from "path";
import rulesSource from "./check_rules";

//#region utils ===================================================================================================== //

declare global {
	interface Array<T> {
		count(value: T): number;
		reversed(): T[];
	}

	interface String {
		removeSuffix(suffix: string): string;
		removeSuffix(suffix: string): string;
		removePrefixSuffix(prefix: string, suffix: string): string;
		reduce<T>(callbackfn: (previousValue: T, currentChar: string, currentIndex: number, string: string) => T,
		          initialValue: T): T;
		reduceRight<T>(callbackfn: (previousValue: T, currentChar: string, currentIndex: number, string: string) => T,
		               initialValue: T): T;
		count(char: string): number;
		reversed(): string;
	}
}

Array.prototype.count = function<T>(this: T[], countValue: T): number {
	return this.reduce((count, currentValue) => {
		return count + (currentValue === countValue ? 1 : 0);
	}, 0);
}
Array.prototype.reversed = function<T>(this: T[]): T[] {
	return this.reduceRight((reversed, value) => {
		reversed.push(value);
		return reversed;
	}, [] as T[]);
}

String.prototype.removeSuffix = function(this: string, suffix: string): string {
	if(this.endsWith(suffix)) {
		return this.substring(0, this.length - suffix.length);
	}
	return this;
}
String.prototype.reduce = function<T>(this: string,
                                      callbackfn: (previousValue: T,
                                                   currentChar: string,
                                                   currentIndex: number,
                                                   string: string) => T,
                                      initialValue: T): T {

	let accumulator = initialValue;

	for(let i = 0; i < this.length; ++i) {
		accumulator = callbackfn(accumulator, this[i], i, this);
	}

	return accumulator;
}
String.prototype.reduceRight = function<T>(this: string,
                                           callbackfn: (previousValue: T,
                                                        currentChar: string,
                                                        currentIndex: number,
                                                        string: string) => T,
                                           initialValue: T): T {

	let accumulator = initialValue;

	for(let i = this.length - 1; i >= 0; --i) {
		accumulator = callbackfn(accumulator, this[i], i, this);
	}

	return accumulator;
}
String.prototype.count = function(this: string, countChar: string): number {
	return this.reduce((count, currentChar) => {
		return count + (currentChar === countChar ? 1 : 0);
	}, 0);
}
String.prototype.reversed = function(this: string): string {
	return this.reduceRight((reversed, char) => {
		return reversed + char;
	}, "");
}

//#endregion

//#region getting the writing and formatting rules ================================================================== //

const rules = Object.entries(rulesSource).reduce((rules, [description, patternOrPatterns]) => {
	rules[description] = (patternOrPatterns instanceof Array ? patternOrPatterns : [patternOrPatterns])
	                     .map((pattern) => {
		                     let flags = pattern.flags;
		                     if(!flags.includes("g")) flags += "g";
		                     return new RegExp(pattern, flags);
	                     });
	return rules;
}, {} as {
	[description: string]: RegExp[];
}) as {
	readonly [description: string]: RegExp[];
};

//#endregion

//#region searching for files ======================================================================================= //

function readdirRecursiveSync(rootPath: PathLike): string[] {
	return readdirSync(rootPath).flatMap((childPath) => {
		childPath = `${rootPath}/${childPath}`;

		if(statSync(childPath).isDirectory()) {
			return readdirRecursiveSync(childPath);
		}

		return [normalizePath(childPath)];
	});
}

const filepaths = readdirRecursiveSync("docs")
	.filter((filepath) => filepath.endsWith(".md"))
	.filter((filepath) => !filepath.endsWith("cc-by-sa-4.0.md"));

//#endregion

//#region searching for problems ==================================================================================== //

enum ProblemOrigin {
	T3M_RULES = "t3m",
	MARKDOWNLINT = "markdownlint"
}

class Problem {
	private constructor(readonly origin: ProblemOrigin,

	                    readonly filepath: string,
	                    readonly description: string,

	                    readonly lines: string[],

	                    readonly beginLineno: number,
	                    readonly beginColumn: number,

	                    readonly endLineno: number,
	                    readonly pastLastColumn: number) {
	}

	private static countLn(str: string): number {
		return str.count("\n");
	}
	private static lineSpan(str: string, beginLineno: number, endLineno: number): string[] {
		return str.split(/\r?\n|\n/).slice(beginLineno - 1, endLineno);
	}

	static fromIndex(origin: ProblemOrigin,
	                 filepath: string,
	                 description: string,
	                 fileContents: string,
	                 index: number,
	                 length: number): Problem {

		const lastLinePattern = /\n(.*)$/;

		const prefix = fileContents.substring(0, index);
		const beginLineno = Problem.countLn(prefix) + 1;
		const beginColumn = (prefix.match(lastLinePattern)?.[1] ?? prefix).length + 1;

		const substring = fileContents.substring(index, index + length);
		const endLineno = Problem.countLn(substring) + beginLineno;

		const match = substring.match(lastLinePattern);
		const pastLastColumn = (match === null
		                        ? beginColumn + substring.length
		                        : match[1].length + 1);

		const lines = Problem.lineSpan(fileContents, beginLineno, endLineno);

		return new Problem(origin, filepath, description, lines, beginLineno, beginColumn, endLineno, pastLastColumn);
	}

	static fromLinenoAndColumn(origin: ProblemOrigin,
	                           filepath: string,
	                           description: string,
	                           fileContents: string,
	                           beginLineno: number,
	                           beginColumn: number,
	                           endLineno: number = beginLineno,
	                           pastLastColumn?: number): Problem {

		const lines = Problem.lineSpan(fileContents, beginLineno, endLineno);

		if(pastLastColumn === undefined) pastLastColumn = lines[endLineno - beginLineno].length + 1;

		return new Problem(origin,
		                   filepath,
		                   description,
		                   lines,
		                   beginLineno,
		                   beginColumn,
		                   endLineno,
		                   pastLastColumn);
	}
}

const problems = filepaths.reduce((problems, filepath) => {
	const file = readFileSync(filepath).toString();

	for(const [description, patterns] of Object.entries(rules)) {
		patterns.forEach((pattern) => {
			for(const match of file.matchAll(pattern)) {
				problems.push(Problem.fromIndex(ProblemOrigin.T3M_RULES,
				                                filepath,
				                                description,
				                                file,
				                                match.index!!,
				                                match[0].length));
			}
		});
	}

	markdownlint.sync({
		files: filepath,
		config: { "line-length": false }
	})[filepath].forEach((result) => {
		let description = `${result.ruleNames.join("/")} ${result.ruleDescription}`;
		if(result.errorDetail) description += ` [${result.errorDetail}]`;

		const [beginColumn, pastLastColumn] = (result.errorRange
		                                       ? [result.errorRange[0], result.errorRange[0] + result.errorRange[1]]
		                                       : [1, undefined]);

		problems.push(Problem.fromLinenoAndColumn(ProblemOrigin.MARKDOWNLINT,
		                                          filepath,
		                                          description,
		                                          file,
		                                          result.lineNumber,
		                                          beginColumn,
		                                          result.lineNumber,
		                                          pastLastColumn));
	});

	return problems;
}, [] as Problem[]);

//#endregion

//#region printing problems

const fileInfoStyle = chalk.bold;
const errorNonWsStyle = chalk.red.bold;
const errorWsStyle = chalk.red.bgRed;

function errorStyle(text: string): string {
	return text.replace(/(\S+)/g, (_, s) => errorNonWsStyle(s))
	           .replace(/(\s+)/g, (_, s) => errorWsStyle(s));
}

function linenoPrefix(lineno: number): string {
	return lineno.toString().padStart(5, " ") + " | ";
}

function printProblem(problem: Problem) {
	const {
		origin,

		filepath,
		description,

		lines,

		beginLineno,
		beginColumn,

		endLineno,
		pastLastColumn: endColumn
	} = problem;

	console.log(`${fileInfoStyle(`${filepath}:${beginLineno}:${beginColumn}:`)} (${origin}) ${description}`);

	if(lines.length === 1) {
		const line = lines[0];

		console.log(linenoPrefix(beginLineno) +
		            line.substring(0, beginColumn - 1) +
		            errorStyle(line.substring(beginColumn - 1, endColumn - 1)) +
		            line.substring(endColumn - 1));
		return;
	}

	lines.forEach((line, i) => {
		switch(i) {
			case(0): {
				console.log(linenoPrefix(beginLineno) +
				            line.substring(0, beginColumn - 1) +
				            errorStyle(line.substring(beginColumn - 1)));
				break;
			}
			case(lines.length - 1): {
				console.log(linenoPrefix(endLineno) +
				            errorStyle(line.substring(0, endColumn - 1)) +
				            line.substring(endColumn - 1));
				break;
			}
			default: {
				console.log(linenoPrefix(beginLineno + i) + errorStyle(line));
				break;
			}
		}
	});
}

if(problems.length !== 0) {
	process.exitCode = 32;

	printProblem(problems[0]);

	problems.slice(1).forEach((problem) => {
		console.log();
		printProblem(problem);
	});
}

//#endregion
